src/Parser/Monk/Brewmaster/Modules/Normalizers/IronskinBrew.js

```diff
@@ -0,0 +1,161 @@
+        isbCasts[sourceID] = isbCasts[sourceID] ? isbCasts[sourceID] : [];
+        isbCasts[sourceID].push(event);
+        isbAbility = isbAbility ? isbAbility : event.ability;
+      } else if((event.type === "applybuff" || event.type === "refreshbuff" || event.type === "removebuff") && SPELLS.IRONSKIN_BREW_BUFF.id === spellId) {
+        isbBuffs[targetID] = isbBuffs[targetID] ? isbBuffs[targetID] : [];
+        isbBuffs[targetID].push(event);
+        isbBuffAbility = isbBuffAbility ? isbBuffAbility : event.ability;
+      }
+    }
+    // these targets are bugged, having >= 1 ISB casts but no ISB buff
+    // events
+    const buggedTargets = Object.keys(isbCasts).filter((target) => !isbBuffs.hasOwnProperty(target));
+    debug && console.log(isbCasts, isbBuffs, buggedTargets);
+    if(buggedTargets.length === 0) {
+      // nobody is bugged, skip the rest
+      return events;
+    }
+
+    if(!isbBuffAbility) {
+      isbBuffAbility = {
+        guid: SPELLS.IRONSKIN_BREW_BUFF.id,
+        icon: SPELLS.IRONSKIN_BREW_BUFF.icon,
+        name: SPELLS.IRONSKIN_BREW_BUFF.name,
+      };
+    }
+    
+    for(let i = 0; i < buggedTargets.length; i++) {
+      const targetID = parseInt(buggedTargets[i], 10);
+      const targetInfo = this._combatantInfoEvents.find(event => event.sourceID === targetID);
+      // potent kick increases the duration of the isb buff
+      const pk = targetInfo.artifact.find(trait => trait.spellID === SPELLS.POTENT_KICK.id);
+      // quick sip causes purifying brew to add 1 second of the
+      // isb buff
+      const qs = targetInfo.artifact.find(trait => trait.spellID === SPELLS.QUICK_SIP.id);
+      const pk_rank = pk ? pk.rank : 0;
+      const has_qs = qs ? qs.rank === 1 : false;
+
+      const DURATION = ISB_BASE_DURATION + pk_rank * PK_EXTRA_DURATION;
+      const CAP = DURATION_CAP_FACTOR * DURATION;
+
+      const fab = (ty, ts) => { 
+        return {
+          timestamp: ts,
+          type: ty,
+          ability: isbBuffAbility,
+          sourceID: targetID,
+          targetID: targetID,
+          sourceIsFriendly: true,
+          targetIsFriendly: true,
+          __fabricated: true,
+          __fabricatedBy: "isb_normalize",
+        };
+      };
+
+      // first we are going to fabricate the initial applybuff
+      // immediately before the first damage the target would take. this
+      // is done because the bug occurs when we have 100% uptime, and if
+      // we don't set the timestamp to be late enough we can actually
+      // get much less than 100% uptime
+      const firstDmg = events.find(event => event.type === "damage" && event.targetID === targetID);
+      const firstDmgIdx = events.indexOf(firstDmg);
+      events.splice(firstDmgIdx, 0, fab("applybuff", firstDmg.timestamp - 1));
+      let currentBuffTime = DURATION;
+      // now we go through the event list and fabricate correct buff
+      // events
+      for(let j = firstDmgIdx; j < events.length; j++) {
```

Comment:
I still don't really understand what this does. If I understand it correctly the only issue you're trying to fix is if ISB has a 100% uptime it never shows, right? So then wouldn't it suffice to insert it at the same spot as ApplyBuff does; after the combatantinfos? Why do you need to removebuff and refresh?

Reply:
The idea had been to reconstruct *each* event that would be present. In theory, this is an initial ApplyBuff, then (I thought) a bunch of RefreshBuff events. The RemoveBuff stuff is present to check when I've gotten it wrong. However, upon reviewing more logs I've realized that the refreshbuff doesn't actually fire for ISB even on not-bugged logs (I expected it to behave like RJW, which has a refreshbuff event at each cast where the buff is already applied).

So, to match actual logs you are correct: I can just add an ApplyBuff immediately after the combatantinfos. This will simplify the module dramatically. I'll update the PR with that this evening.